New Post for Angular Image Directive (#3544)
GoogleChrome/developer.chrome.com: site/en/blog/angular-image-directive/index.md

The Aurora team believes in leveraging the power of frameworks to provide baked-in solutions to common developer challenges. Their first foray into the image optimization space was the [Next.js image component](https://web.dev/image-component/). They considered this component to be a  testing ground for whether improving the developer experience (DX) of image optimization could lead to performance wins for more apps using frameworks.

The first set of results from Next.js user [Leboncoin](https://medium.com/leboncoin-engineering-blog/how-we-are-improving-our-web-performance-9f850d59d810) was encouraging. Leboncoin saw a significant LCP improvement (from 2.4s to 1.7s) after they started using `next/image`. The subsequent adoption of `next/image` in the community played a role in the increase of Next.js origins meeting LCP thresholds. Soon there were [requests](https://github.com/angular/angular/issues/42765) for similar features in other frameworks, one of them being [Angular](https://angular.io/).  

> As a result, Aurora consulted with [Angular](https://github.com/NG-ZORRO/ng-zorro-antd/issues/6514) and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to bring image optimization defaults to Angular. 

Reviewer: Can we remove the ng-zorro link?

Reviewer 2: done

> As a result, Aurora consulted with [Angular](https://github.com/NG-ZORRO/ng-zorro-antd/issues/6514) and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to bring image optimization defaults to Angular. 

Reviewer 2: ```suggestion
As a result, Aurora consulted with Angular and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to bring image optimization defaults to Angular. 
```

**Website 1:** Uses native `<img>` elements with images served through the Imgix CDN (with default configuration options)

**Website 2:** Use the image directive for all images. It also includes the optimizations recommended directly by warnings or errors thrown by the directive.

<figure>
{% Img src="image/IypihH3o5cSpEMVp5i08dp69otp2/Jc3mkO2gYZGOPul8Dpgk.png", alt="Filmstrip comparison: Website 1 with native image tags Vs. Website 2 with the Angular image directive.", width="800", height="391" %}
</figure>

{% Aside 'success' %}
When running these demos through [WebPageTest](https://webpagetest.org) lab tests on a physical Moto G4 device, there is a substantial improvement of 56% in LCP. The original LCP of 4.5 seconds on Website 1 using native image tags reduces to 2 seconds on Website 2 using the image directive. 
{% endAside %}

The team worked with partners to validate the image directive's performance impact on real enterprise Angular applications. 

One of these partners was [Land's End](https://www.landsend.com/). It was expected that their site would be a good test case for results that real applications might see.

> [Lighthouse lab testing](https://philipwalton.com/articles/my-challenge-to-the-web-performance-community/) was performed on one of their routes before and after using the image directive. On desktop, their median LCP decreased from 12.0s to 3.0s, a 75% improvement in LCP. On mobile, the median LCP decreased from 20.2s to 12.0s (40.6% improvement). 

Reviewer: Land's End asked for this to be clarified.

```suggestion
[Lighthouse lab testing](https://philipwalton.com/articles/my-challenge-to-the-web-performance-community/) was performed on their QA environment before and after using the image directive. On desktop, their median LCP decreased from 12.0s to 3.0s, a 75% improvement in LCP. On mobile, the median LCP decreased from 20.2s to 12.0s (40.6% improvement). 
```

---
layout: 'layouts/blog-post.njk'
title: Optimizing Images with the Angular Image Directive
description: 
  Introducing the new Angular image directive (NgOptimizedImage) that brings built-in performance optimization techniques to images in Angular apps.
date: 2022-08-26
authors:
  - karaerickson
  - leenasohoni
hero: 'image/IypihH3o5cSpEMVp5i08dp69otp2/ZKPh5cWm0KrZ9XYK7UDV.jpg'
alt: >
  Two people looking at a website with images displayed.
tags:
  - performance

>   - news

Reviewer 2: ```suggestion
  - news
  - aurora-project
```

---

> In May 2022, the [Aurora](https://web.dev/introducing-aurora/) and Angular teams announced that they would collaborate on an [image directive](https://angular.io/guide/roadmap#improve-image-performance) for Angular. The directive was recently released for developer preview as part of Angular v14.2! This post talks about how the new image directive, `NgOptimizedImage`, supports image optimization in Angular.

Reviewer 2: ```suggestion
In May 2022, the [Aurora](https://web.dev/introducing-aurora/) and Angular teams announced that they would collaborate on an [image directive](https://angular.io/guide/roadmap#improve-image-performance) for Angular. The directive was recently released for developer preview as part of Angular v14.2. This post talks about how the new image directive, `NgOptimizedImage`, supports image optimization in Angular.
```

## Background

> Images are a ubiquitous and crucial component of web user experience, with [99.9%](https://almanac.httparchive.org/en/2021/media#images) of web pages generating requests for one or more images. Images are also the most significant contributors to page weight, constituting a median of [982 kilobytes](https://almanac.httparchive.org/en/2021/page-weight#fig-2) per page.

Reviewer 2: ```suggestion
Images are a common and crucial component of web user experience, with [99.9%](https://almanac.httparchive.org/en/2021/media#images) of web pages generating requests for one or more images. Images are also the most significant contributors to page weight, constituting a median of [982 kilobytes](https://almanac.httparchive.org/en/2021/page-weight#fig-2) per page.
```

Due to their growing number and size, images can hinder the performance of web pages and affect [Core Web Vitals](https://web.dev/vitals/) metrics. For [79.4% of desktop pages](https://almanac.httparchive.org/en/2021/media#fig-4), an image was the Largest Contentful Paint ([LCP](https://web.dev/lcp/)) element in 2021. The pursuit of optimized images has thus become a constant endeavor for many of us.

The Aurora team believes in leveraging the power of frameworks to provide baked-in solutions to common developer challenges. Their first foray into the image optimization space was the [Next.js image component](https://web.dev/image-component/). They considered this component to be a  testing ground for whether improving the developer experience (DX) of image optimization could lead to performance wins for more apps using frameworks.

> The first set of results from Next.js user [Leboncoin](https://medium.com/leboncoin-engineering-blog/how-we-are-improving-our-web-performance-9f850d59d810) was encouraging. Leboncoin saw a significant LCP improvement (from 2.4s to 1.7s) after they started using `next/image`. The subsequent adoption of next/image in the community played a role in the increase of Next.js origins meeting LCP thresholds. Soon there were [requests](https://github.com/angular/angular/issues/42765) for similar features in other frameworks, one of them being [Angular](https://angular.io/).  

Reviewer 2: ```suggestion
The first set of results from Next.js user [Leboncoin](https://medium.com/leboncoin-engineering-blog/how-we-are-improving-our-web-performance-9f850d59d810) was encouraging. Leboncoin saw a significant LCP improvement (from 2.4s to 1.7s) after they started using `next/image`. The subsequent adoption of `next/image` in the community played a role in the increase of Next.js origins meeting LCP thresholds. Soon there were [requests](https://github.com/angular/angular/issues/42765) for similar features in other frameworks, one of them being [Angular](https://angular.io/).  
```

> As a result, Aurora consulted with [Angular](https://github.com/NG-ZORRO/ng-zorro-antd/issues/6514) and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to h bring image optimization defaults to Angular. 

Reviewer 2: ```suggestion
As a result, Aurora consulted with [Angular](https://github.com/NG-ZORRO/ng-zorro-antd/issues/6514) and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to bring image optimization defaults to Angular. 
```

> As a result, Aurora consulted with [Angular](https://github.com/NG-ZORRO/ng-zorro-antd/issues/6514.) and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to h bring image optimization defaults to Angular. 

Reviewer 3: ```suggestion
As a result, Aurora consulted with [Angular](https://github.com/NG-ZORRO/ng-zorro-antd/issues/6514) and [Nuxt](https://nuxtjs.org/) to prototype image components for these frameworks. The [Nuxt image component](https://image.nuxtjs.org/) was released last year. Now the Angular image directive (`NgOptimizedImage`) has been released to h bring image optimization defaults to Angular. 
```

## Opportunity

Angular is one of the leading JavaScript frameworks used by developers today. It is used by more than [50k of the origins](https://datastudio.google.com/u/0/reporting/55bc8fad-44c2-4280-aa0b-5f3f0cd3d2be/page/M6ZPC?s=o6zLzlTpWaI&params=%7B%22df44%22:%22include%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580Angular%25EE%2580%2580Next.js%25EE%2580%2580Nuxt.js%22,%22df46%22:%22include%25EE%2580%25800%25EE%2580%2580IN%25EE%2580%2580mobile%22%7D) crawled by [HTTPArchive](https://httparchive.org/) on mobile and boasts [nearly 3 million](https://www.npmjs.com/package/@angular/core) weekly downloads on NPM. 

<figure>

> {% Img src="image/IypihH3o5cSpEMVp5i08dp69otp2/epoxZjv8snr3xdEKl9UB.png", alt="LCP for Angular websites over the last one year", width="800", height="711" %}

Reviewer 2: ```suggestion
{% Img src="image/IypihH3o5cSpEMVp5i08dp69otp2/epoxZjv8snr3xdEKl9UB.png", alt="LCP for Angular websites over the last one year.", width="800", height="711" %}
```

</figure>

Looking at the Core Web Vitals scores, the percentage of Angular origins that meet "good" LCP thresholds still needs work. Only 18.74% of Angular sites had good LCP on mobile in June 2022. Since images are the LCP element for more than 70% of the web pages on mobile and desktop, unoptimized LCP images could be one of the primary causes of poorer LCP on Angular websites. 

> The Angular image directive was designed to l help improve these numbers. 

Reviewer 2: ```suggestion
The Angular image directive was designed to help improve these numbers. 
```

## MVP for the NgOptimizedImage directive

The MVP of the Angular image directive builds on lessons from the image components Aurora has built to date while adapting the design to Angular’s client-side rendering experience. Many of the standard image optimization problems have been addressed by either:

> * Providing strong defaults or 
> * Throwing errors/warnings to ensure conformance to best practices.

Reviewer 2: ```suggestion
* Providing strong defaults.
* Throwing errors or warnings to ensure conformance to best practices.
```

The highlights of the design are as follows:

1. **Intelligent lazy loading**

>     Images that are invisible to the user on page load (e.g., below-the-fold images or hidden carousel images) should ideally be [lazy-loaded](https://web.dev/lazy-loading-images/). Lazy loading frees up browser resources to load other critical text, media, or scripts. Most images are non-critical and should be lazy-loaded, but only [7.8% of pages](https://almanac.httparchive.org/en/2021/resource-hints#fig-16) used native lazy loading in 2021.

Reviewer 2: ```suggestion
    Images that are invisible to the user on page load (for example, below-the-fold images or hidden carousel images) should ideally be [lazy-loaded](https://web.dev/lazy-loading-images/). Lazy loading frees up browser resources to load other critical text, media, or scripts. Most images are non-critical and should be lazy-loaded, but only [7.8% of pages](https://almanac.httparchive.org/en/2021/resource-hints#fig-16) used native lazy loading in 2021.
```

    The Angular image directive lazy loads non-critical images by default and only eagerly loads images specially marked as `priority`. This ensures that most images exhibit optimal loading behavior.

2. **Prioritization of critical images**

    Adding resource hints (e.g., <code>[preload](https://web.dev/preload-critical-assets/)</code> or <code>[preconnect](https://web.dev/uses-rel-preconnect/)</code>) to prioritize the loading of critical images is a [recommended best practice](https://web.dev/preload-critical-assets/). However, most apps are not using them. According to the 2021 Web Almanac, only [12.7% of mobile pages](https://almanac.httparchive.org/en/2021/resource-hints#fig-2) use preconnect hints and [only 22.1% of mobile pages](https://almanac.httparchive.org/en/2021/resource-hints#fig-2) use preload hints.

	The image directive acts on two fronts when images are marked as priority.

>     * It sets the <code>[fetchpriority](https://web.dev/priority-hints/#increase-the-priority-of-the-lcp-image)</code> of the image to <code>"high"</code> so that the browser knows that it should download the image with a high priority.

Reviewer 2: ```suggestion
    * It sets the `[fetchpriority](https://web.dev/priority-hints/#increase-the-priority-of-the-lcp-image)` of the image to `"high"` so that the browser knows that it should download the image with a high priority.
```

>     * In development mode, a runtime check confirms that a <code>preconnect</code> resource hint has been included corresponding to the image's origin.

Reviewer 2: ```suggestion
    * In development mode, a runtime check confirms that a `preconnect` resource hint has been included corresponding to the image's origin.
```

    In development mode, the directive also uses the [PerformanceObserver API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver) to verify that the [LCP image](https://developer.mozilla.org/en-US/docs/Web/API/LargestContentfulPaint/element) has been marked `priority` as expected. If it's not marked `priority`, an error is thrown, instructing the developer to add the `priority` attribute to the LCP image.

    Ultimately, this combination of automation and conformance ensures that the LCP image has a `preconnect` hint, a `fetchpriority` attribute value of `high`, and isn't lazy loaded.

3. **Optimized configuration for popular image tooling**

>     It's recommended that Angular applications [leverage image CDNs](https://web.dev/image-cdns/), which often provide optimization services by default.

Reviewer 2: ```suggestion
    It's recommended that Angular applications [use image CDNs](https://web.dev/image-cdns/), which often provide optimization services by default.
```

    
    2. **Aspect ratio:** The image directive throws an error to let developers know if the aspect ratio of the `width`:`height` defined in the HTML is not close to the actual aspect ratio of the rendered image. This can cause the image to look distorted on screen. This can happen if 
        1. You have defined the wrong dimensions (width or height) by mistake or
        2. If you have defined one dimension by percentage in your CSS, but not the other (for example, `width: 100%` needs `height: auto` to ensure the image grows in both dimensions). 
   
    3. **Oversized images:** If the image does not define a `srcset `and the intrinsic image is significantly larger than the rendered image, the directive will display a warning suggesting the use of the `srcset` and `sizes` attributes.
    

>     4. **Image density:** The directive will throw an error if you try to include an image in the `srcset` with a pixel density of more than `3x`. Descriptors higher than `2x` are generally not recommended because it has the unintended consequence of forcing high-resolution mobile devices to download huge images. Moreover, the human eye can't actually tell much of a difference [above 2x](https://web.dev/uses-rel-preconnect/).

Reviewer 3: > Moreover, the human eye can't actually tell much of a difference [above 2x](https://web.dev/uses-rel-preconnect/).

hmm that link doesn't seem right.

Author: There was some discussion about this. Have changed it to another link thats more relevant.

Some of the challenges encountered are as follows:

1. **Supporting resource hints**

    [Preloading critical assets](https://web.dev/preload-critical-assets/) helps the browser discover them earlier. However, including resource hints in Angular apps is complicated because:

    **Manual Addition**: It's difficult for developers to add the `preload` resource hint manually. Angular uses one shared index.html file for the entire project or for all routes in the website. Thus, the `<head>` of the document is the same for every route (at least at serve time). Adding any `preload` hint to the `<head>` would mean that the resource would be preloaded for all routes even where it is not required. Thus, the manual addition of `preload` hints is not recommended.

    **Automatic addition during render:** Utilizing the framework to add preload hints to the head of the document during render in a CSR app does not help. Since rendering occurs after JavaScript is downloaded and executed, the `<head>` will be rendered too late to be of any value. 

>     For the first version of the directive, a combination of `preconnect` hints and priority hints serve to prioritize the image in lieu of a `preload`. However, Aurora is currently working with the Angular CLI team to enable automatic injection of resource hints at build time — stay tuned!

Reviewer 2: ```suggestion
    For the first version of the directive, a combination of `preconnect` hints and priority hints serve to prioritize the image in lieu of a `preload`. However, Aurora is currently working with the Angular CLI team to enable automatic injection of resource hints at build time—stay tuned!
```

2. **Optimizing image size and format on the server**

    As Angular apps are typically client-side rendered, images on the file system cannot be compressed at request time and are served as is. For this reason, using image CDNs is recommended to compress images and convert them into [modern formats like WebP](https://web.dev/serve-images-webp/) or AVIF on demand.

    While the directive does not enforce the use of image CDN’s, it's strongly encouraged to use them with the directive and its built-in loaders ensure that the correct configuration options are used.

## Impact

The following demo demonstrates the difference the Angular image directive can make to image performance. It compares two websites:

> **Website 1:** Uses native `<img>` elements with images served through the Imgix CDN (with default configuration options)

Reviewer 2: ```suggestion
**Website One:** Uses native `<img>` elements with images served through the Imgix CDN (with default configuration options).
```

> **Website 2:** Use the image directive for all images. It also includes the optimizations recommended directly by warnings or errors thrown by the directive.

Reviewer 2: ```suggestion
**Website Two:** Use the image directive for all images. It also includes the optimizations recommended directly by warnings or errors thrown by the directive.
```

<figure>

> {% Img src="image/IypihH3o5cSpEMVp5i08dp69otp2/Jc3mkO2gYZGOPul8Dpgk.png", alt="Filmstrip comparison: Website 1 with native image tags Vs. Website 2 with the Angular image directive.", width="800", height="391" %}

Reviewer 2: ```suggestion
{% Img src="image/IypihH3o5cSpEMVp5i08dp69otp2/Jc3mkO2gYZGOPul8Dpgk.png", alt="Filmstrip comparison: Website One with native image tags versus Website Two with the Angular image directive.", width="800", height="391" %}
```

</figure>

{% Aside 'success' %}

> When running these demos through [WebPageTest](https://webpagetest.org) lab tests on a physical Moto G4 device, there is a substantial improvement of 56% in LCP. The original LCP of 4.5 seconds on Website 1 using native image tags reduces to 2 seconds on Website 2 using the image directive. 

Reviewer 2: ```suggestion
When running these demos through [WebPageTest](https://webpagetest.org) lab tests on a physical Moto G4 device, there is a substantial improvement of 56% in LCP. The original LCP of 4.5 seconds on Website One using native image tags reduces to 2 seconds on Website Two using the image directive. 
```
